torch/fx/passes/backends/cudagraph_partition.py

```diff
@@ -0,0 +1,232 @@
+                    # across layers will make the cudagraph capture very slow.
+                    # therefore, we only run gc for the first graph,
+                    # and disable gc for the rest of the graphs.
+                    stack.enter_context(patch("gc.collect", lambda: None))
+                    stack.enter_context(patch("torch.cuda.empty_cache", lambda: None))
+
+                with torch.cuda.graph(self.cudagraph, pool=self.graph_pool):
+                    # `output` is managed by pytorch's cudagraph pool
+                    # TODO: use weak ref for output to reuse memory
+                    self.output = self.runnable(*args, **kwargs)
+
+        self.cudagraph.replay()
+        return self.output
+
+
+# Part 3: Patch gm with CG Wrapper
+class PiecewiseCompileInterpreter(torch.fx.Interpreter):
+    def __init__(
+        self,
+        module: torch.fx.GraphModule,
+        submod_names_to_cudagraph: list[str],
+    ):
+        super().__init__(module)
+        self.submod_names_to_cudagraph = submod_names_to_cudagraph
+
+    def call_module(
+        self,
+        target: torch.fx.node.Target,
+        args: tuple[torch.fx.node.Argument, ...],
+        kwargs: dict[str, Any],
+    ) -> Any:
+        assert isinstance(target, str)
+        output = super().call_module(target, args, kwargs)
+
+        if target in self.submod_names_to_cudagraph:
+            index = self.submod_names_to_cudagraph.index(target)
+            is_first_graph = index == 0
+            submod = self.fetch_attr(target)
+            self.module.__dict__[target] = CUDAGraphWrapper(
+                submod, gc_disable=not is_first_graph
+            )
+
+        return output
+
+
+def cudagraph_partition_pass(
+    gm: fx.GraphModule,
+    example_inputs: list[torch.Tensor],
+    input_clone_indices: list[int],
+    split_ops: list[str],
+) -> fx.GraphModule:
+    """
+    Partition the graph into subgraphs and wrap them with CUDAGraphWrapper.
+
+    @param gm: the graph to be partitioned
+    @param split_ops: a list of ops to be partitioned
+    @param input_clone_indices: a list of indices of the inputs that need to be cloned
+        to static tensors. This is needed since CUDAGraph requires static tensors.
+        This includes user inputs. This does not include parameters and buffers,
+        which are already static tensors.
+
+    @return: the partitioned graph
+    """
+    # 1. Split graph
+    resolved_split_ops = resolve_defined_ops(split_ops)
+    split_gm, split_items = split_graph(gm, resolved_split_ops)
+    submod_names_to_cudagraph = [
+        item.submod_name for item in split_items if not item.is_cudagraph_unsafe
+    ]
+
+    # split_gm.input_buffers refers to the tensors in the list, instead of the list itself
+    split_gm.input_buffers = [t for t in example_inputs]
+
+    # 2. Wrap submodules with CUDAGraphWrapper
+    PiecewiseCompileInterpreter(split_gm, submod_names_to_cudagraph).run(
```

Comment:
when do we warm up cuda graph?

Reply:
I guess the questions is, this is JIT-style cudagraph, I imagine it also could be a AOT cudagraph, where you warm up with example inputs.